### PR TITLE
deprecate global and process verbosity

### PR DESCRIPTION
**Pull request by tomerd**
motivation: fine grained control over verbosity setting and handling of logging

changes:
* deprecate global verbosity flag
* deprecate Process verbosity attribute in favor of an optional logging handler (closure)

### CODE REVIEW

**Comment by tomerd** [comment]
afaict this is only used by SwiftPM, which is addressed in https://github.com/apple/swift-package-manager/pull/4109

File: Sources/TSCBasic/Process.swift [hunk 1]
```diff
@@ -320,15 +356,36 @@ public final class Process {
...
         self.environment = environment
         self.workingDirectory = nil
         self.outputRedirection = outputRedirection
-        self.verbose = verbose
         self.startNewProcessGroup = startNewProcessGroup
+        self.loggingHandler = loggingHandler
+    }
+
+    @_disfavoredOverload
+    @available(*, deprecated, message: "user version without verbosity flag")
```
Comment (cltnschlosser) [inline Sources/TSCBasic/Process.swift:371]: typo
```suggestion
    @available(*, deprecated, message: "use version without verbosity flag")
```

**Comment by tomerd** [comment]
https://github.com/apple/swift-package-manager/pull/4109

@swift-ci please test

**Comment by tomerd** [comment]
https://github.com/apple/swift-package-manager/pull/4109

@swift-ci please test

**Comment by tomerd** [comment]
https://github.com/apple/swift-package-manager/pull/4109

@swift-ci please test

**Comment by tomerd** [comment]
https://github.com/apple/swift-package-manager/pull/4109

@swift-ci please test

**Review by abertelrud — APPROVED** [review]

File: Sources/TSCBasic/Process.swift [hunk 2]
```diff
@@ -393,9 +450,10 @@ public final class Process {
         }
 
         // Print the arguments if we are verbose.
-        if self.verbose {
-            stdoutStream <<< arguments.map({ $0.spm_shellEscaped() }).joined(separator: " ") <<< "\n"
-            stdoutStream.flush()
+        if let loggingHandler = self.loggingHandler {
+            loggingHandler(arguments.map({ $0.spm_shellEscaped() }).joined(separator: " "))
```
Comment (yim-lee) [inline Sources/TSCBasic/Process.swift:454]: `\n` not needed?

**Review by yim-lee — APPROVED** [review]

File: Sources/TSCBasic/Process.swift (hunk 2 above)
Comment (tomerd) [reply Sources/TSCBasic/Process.swift:454]: not any more, it was for the stream itself. the handler will decide what to do with it now

File: Sources/TSCBasic/Process.swift [hunk 3]
```diff
@@ -205,10 +205,20 @@ public final class Process {
     public typealias OutputClosure = ([UInt8]) -> Void
 
     /// Global default setting for verbose.
-    public static var verbose = false
+    @available(*, deprecated)
+    public static var verbose: Bool {
+        Self.loggingHandler != nil
+    }
+
+    public static var loggingHandler: ((String) -> Void)? = nil
```
[the remark this answers is not in the document]
Comment (weissi) [reply Sources/TSCBasic/Process.swift:213]: also thread unsafe

[the remark this answers is not in the document]
Comment (tomerd) [reply Sources/TSCBasic/Process.swift:213]: yea i think i am going to just remove it and let the clients deal with setting a handler at the call site level. they can always create their own wrapper

[the remark this answers is not in the document]
Comment (tomerd) [reply Sources/TSCBasic/Process.swift:213]: yea, I dont like this global pattern anyway

[the remark this answers is not in the document]
Comment (tomerd) [reply Sources/TSCBasic/Process.swift]: ugh. its quite a bit of a hassle unfortunately. I added lock for now :/ ptal

**Comment by tomerd** [comment]
@swift-ci please test

File: Sources/TSCBasic/Process.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (weissi) [reply Sources/TSCBasic/Process.swift]: lgtm, thanks!

**Pull request merged**